### ISSUE TITLE
B018 false positive for notebooks

### ISSUE THREAD

**Issue by david-waterworth** [astral-sh/ruff#6586]
Commonly with notebooks the `result` of a calculation is often implicitly printed, i.e.

```
x = 1 + 2
x
```

I believe x.\_\_repl\_\_() is called.

`ruff` reports `B018: useless-expression`. I'm not sure this is easy to detect as it doesn't have to be the last statement of a cell - i.e. the output of the following is to print x and ignore y.

```
x = 1 + 2
x
y = x + 2
y
```

**Comment by charliermarsh** [comment]
I think this is a case where I'd likely recommend disabling the rule for notebooks. @dhruvmanila, any idea if we could detect whether the expression is the last line in a cell? I know we have the splits for import sorting, but not sure how accessible they are...

**Comment by dhruvmanila** [comment]
I would recommend disabling as well. Otherwise, what could be done is using the [`cell_offsets`](https://github.com/astral-sh/ruff/blob/81b1176f9928ca146f4f6a3ab32804e5fcc00c75/crates/ruff/src/jupyter/notebook.rs#L403-L406) (aka the cell boundaries) which can be used to detect the last non-empty line of the cell. This would be limited as pointed out by @david-waterworth in the second example where these expressions are present in the middle of the cell.

**Comment by charliermarsh** [comment]
Given:

```python
x = 1 + 2
x
y = x + 2
y
```

I would assume we want to flag `x` (second line) as useless but not `y` (fourth line), but it sounds like neither should be flagged. Why is that?

**Comment by dhruvmanila** [comment]
Oh yes, you're correct. I just verified that the value of `x` won't be printed. So, I think it's safe to avoid checking for unused expressions for the last non-empty line i.e., basically if there's an expression before the cell boundary avoid flagging `B018` for it.

**Comment by charliermarsh** [comment]
@david-waterworth - Does that make sense to you? Your initial message said "it doesn't have to be the last statement of a cell".

**Comment by david-waterworth** [comment]
@charliermarsh yes you're correct - I thought it printed the first statement but I checked and it prints the last.

Also

```
x = 1
x

for i in range(10):
    i
```

Doesn't print anything so not really sure how exactly it works and I've not been able to find a definitive description. But it seems to print the value of the last statement but only if it's not indented.

**Comment by dhruvmanila** [comment]
> Doesn't print anything so not really sure how exactly it works and I've not been able to find a definitive description. But it seems to print the value of the last statement but only if it's not indented.

I think it's because the entire `for` loop is a single statement in the AST. The `i` expression is part of the top level `for` loop (child node):
```python
for i in range(10):
    i
```

while in the previous example, the last node is an expression:
```python
x
```

Expressions evaluate to a value while statements do not and that's why the second example prints the value of `x` while the first example (`for` loop) doesn't.

**Comment by dhruvmanila** [comment]
We might want to ignore `B015` (useless-comparison) as well.

**Comment by dhruvmanila** [comment]
(Merging this with a general tracking issue: https://github.com/astral-sh/ruff/issues/8669)
